src/core/ddsi/src/q_plist.c

```diff
@@ -2760,6 +2760,55 @@ dds_qos_t * nn_xqos_dup (const dds_qos_t *src)
   return dst;
 }
 
+#ifdef DDSI_INCLUDE_SECURITY
+static void fill_property(dds_property_t *prop, const char *name, const char *value)
+{
+  prop->name = ddsrt_strdup(name);
+  prop->value = ddsrt_strdup(value);
+  prop->propagate = false;
+}
+
+void nn_xqos_mergein_security_config (dds_qos_t *xqos, const struct omg_security_configuration_type *cfg)
+{
+  if (cfg)
+  {
+    /* Add DDS Security configuration to the QoS as a Property policy,
+     * used by the security plugins to get their proper settings. */
+    dds_property_t *properties;
+    uint32_t idx = 0;
+
+    /* No property should be present yet. */
+    assert(!(xqos->present & QP_PROPERTY_LIST));
+
+    /* Prepare properties. */
+    properties = ddsrt_malloc(8 /* max */ * sizeof (dds_property_t));
+
+    /* Fill properties. */
+    fill_property(&(properties[idx++]), "dds.sec.auth.identity_ca", cfg->authentication_properties.identity_ca);
+    fill_property(&(properties[idx++]), "dds.sec.auth.private_key", cfg->authentication_properties.private_key);
+    fill_property(&(properties[idx++]), "dds.sec.auth.identity_certificate", cfg->authentication_properties.identity_certificate);
+    fill_property(&(properties[idx++]), "dds.sec.access.permissions_ca", cfg->access_control_properties.permissions_ca);
+    fill_property(&(properties[idx++]), "dds.sec.access.governance", cfg->access_control_properties.governance);
+    fill_property(&(properties[idx++]), "dds.sec.access.permissions", cfg->access_control_properties.permissions);
+    if (cfg->authentication_properties.password && (strlen(cfg->authentication_properties.password) != 0))
+    {
+      fill_property(&(properties[idx++]), "dds.sec.auth.password", cfg->authentication_properties.password);
+    }
+    if (cfg->authentication_properties.trusted_ca_dir && (strlen(cfg->authentication_properties.trusted_ca_dir) != 0))
+    {
+      fill_property(&(properties[idx++]), "dds.sec.auth.trusted_ca_dir", cfg->authentication_properties.trusted_ca_dir);
+    }
+
+    /* Add properties. */
+    xqos->present |= QP_PROPERTY_LIST;
+    xqos->property.value.n = idx;
+    xqos->property.value.props = properties;
```

Comment:
There is a leak here if the property list is not empty.
We are not merging the properties, we are replacing the properties here.

Reply:
@kurtuluso this issue is addressed in PR #278